File: examples/vslice/src/test/scala/csw/examples/vslice/TestEnv.scala
package csw.examples.vslice

import java.io.File

import akka.actor.ActorSystem
import akka.util.Timeout
import com.typesafe.config.ConfigFactory
import csw.examples.vslice.assembly.TromboneAssembly
import csw.examples.vslice.hcd.TromboneHCD
import csw.services.alarms.{AlarmService, AlarmServiceAdmin}
import csw.services.cs.akka.ConfigServiceClient
import csw.services.events.{EventService, EventServiceAdmin}

import scala.concurrent.Await
import scala.concurrent.duration._

/**
 * Helper class for setting up the test environment
 */
object TestEnv {

  // For the tests, store the HCD's configuration in the config service (Normally, it would already be there)
  def createTromboneHcdConfig()(implicit system: ActorSystem): Unit = {
    val config = ConfigFactory.parseResources(TromboneHCD.resource.getPath)
    implicit val timeout = Timeout(5.seconds)
    Await.ready(ConfigServiceClient.saveConfigToConfigService(TromboneHCD.tromboneConfigFile, config), timeout.duration)
  }

  // For the tests, store the assembly's configuration in the config service (Normally, it would already be there)
  def createTromboneAssemblyConfig()(implicit system: ActorSystem): Unit = {
    createTromboneHcdConfig()
    implicit val timeout = Timeout(5.seconds)
    val config = ConfigFactory.parseResources(TromboneAssembly.resource.getPath)
    Await.ready(ConfigServiceClient.saveConfigToConfigService(TromboneAssembly.tromboneConfigFile, config), 5.seconds)
  }

  // Reset all redis based services before a test (assumes they are sharing the same Redis instance)
  def resetRedisServices()(implicit system: ActorSystem): Unit = {
    import system.dispatcher
    implicit val timeout = Timeout(10.seconds)
    // clear redis
    val eventService = Await.result(EventService(), timeout.duration)
    val eventServiceAdmin = EventServiceAdmin(eventService)
    Await.ready(eventServiceAdmin.reset(), timeout.duration)

    // initialize the list of alarms
    val alarmService = Await.result(AlarmService(), timeout.duration)
    val alarmServiceAdmin = AlarmServiceAdmin(alarmService)
    //$CSW_INSTALL/conf/alarms.conf
    val cswInstall = Option(System.getenv("CSW_INSTALL"))
    if (cswInstall.isEmpty) {
      println("Environment variable CSW_INSTALL is not set")
      System.exit(1)
    }
    Await.ready(alarmServiceAdmin.initAlarms(new File(s"${cswInstall.get}/conf/alarms.conf")), timeout.duration)
  }
}

File: examples/vsliceJava/src/test/java/csw/examples/vsliceJava/TestEnv.java
package csw.examples.vsliceJava;

import akka.actor.ActorSystem;
import akka.util.Timeout;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import csw.examples.vsliceJava.assembly.TromboneAssembly;
import csw.examples.vsliceJava.hcd.TromboneHCD;
import javacsw.services.alarms.IAlarmService;
import javacsw.services.alarms.IAlarmServiceAdmin;
import javacsw.services.alarms.JAlarmService;
import javacsw.services.alarms.JAlarmServiceAdmin;
import javacsw.services.cs.akka.JConfigServiceClient;
import javacsw.services.events.IEventService;
import javacsw.services.events.IEventServiceAdmin;
import javacsw.services.events.JEventService;
import javacsw.services.events.JEventServiceAdmin;
import scala.concurrent.duration.FiniteDuration;

import java.io.File;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Helper class for setting up the test environment
 */
@SuppressWarnings("WeakerAccess")
public class TestEnv {

  // For the tests, store the HCD's configuration in the config service (Normally, it would already be there)
  public static void createTromboneHcdConfig(ActorSystem system) throws ExecutionException, InterruptedException {
    Config config = ConfigFactory.parseResources(TromboneHCD.resource.getPath());
    Timeout timeout = new Timeout(5, TimeUnit.SECONDS);
    JConfigServiceClient.saveConfigToConfigService(TromboneHCD.tromboneConfigFile, config, system, timeout).get();
  }

  // For the tests, store the assembly's configuration in the config service (Normally, it would already be there)
  public static void createTromboneAssemblyConfig(ActorSystem system) throws ExecutionException, InterruptedException {
    createTromboneHcdConfig(system);
    Timeout timeout = new Timeout(5, TimeUnit.SECONDS);
    Config config = ConfigFactory.parseResources(TromboneAssembly.resource.getPath());
    JConfigServiceClient.saveConfigToConfigService(TromboneAssembly.tromboneConfigFile, config, system, timeout).get();
  }

  // Reset all redis based services before a test (assumes they are sharing the same Redis instance)
  public static void resetRedisServices(ActorSystem system) throws Exception {
    int t = 10;
    TimeUnit u = TimeUnit.SECONDS;
    Timeout timeout = Timeout.durationToTimeout(FiniteDuration.apply(t, u));
    // clear redis
    IEventService eventService = IEventService.getEventService(IEventService.defaultName, system, timeout).get(t, u);
    IEventServiceAdmin eventServiceAdmin = new JEventServiceAdmin(eventService, system);
    eventServiceAdmin.reset().get(t, u);

    // initialize the list of alarms
    IAlarmService alarmService = IAlarmService.getAlarmService(system, timeout).get(t, u);
    IAlarmServiceAdmin alarmServiceAdmin = new JAlarmServiceAdmin(alarmService, system);
    //$CSW_INSTALL/conf/alarms.conf
    String cswInstall = System.getenv("CSW_INSTALL");
    if (cswInstall == null) {
      System.out.println("Environment variable CSW_INSTALL is not set");
      System.exit(1);
    }
    alarmServiceAdmin.initAlarms(new File(cswInstall + "/conf/alarms.conf"), false);
  }
}
